Add new environment variables for experimental/stable pattern

lranjbar opened the pull request:
This PR makes some changes in various places to set up the experimental/stable pattern inside edgeXBuildGoApp to use for git-semver.

The changes I made are:
1) Added two environment variables; `BUILD_EXPERIMENTAL_DOCKER_IMAGE` and `BUILD_STABLE_DOCKER_IMAGE`  into edgeXBuildGoApp to control the building of the `experimental` and `stable` tagged images.
2) Updated edgeXBuildGoApp to use the env.VARIABLE pattern since I was making edits to it.
3) Exposed two previously local variables in edgeXSemver as parameters to be passed in. The defaults chosen were such that the current functionality stays the same.
4) Added utility functions in edgeX.groovy to help the control flow on this.

The overall control flow is:
- When `BUILD_EXPERIMENTAL_DOCKER_IMAGE` is true we will build a new docker image with that tag (docker). Then update an `experimental` tag on Github.
- Right now if you set `BUILD_STABLE_DOCKER_IMAGE` to true it will build the docker image with that tag. However, it might make more sense to have this controlled by `cd-management` so this is still WIP on this part. The two utility functions added in edgeX.groovy are to help out the control flow with this. The current running idea is that this environment variable should be set to true when the commit message follows the pattern `stage(..): ...` in `cd-management`.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/edgexfoundry/ci-management/blob/master/.github/CONTRIBUTING.md
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Added labels
## PR Type
What kind of change does this PR introduce?
<!-- Please check the one that applies to this PR using "x". -->

- [X] Bugfix
- [X] Feature
- [X] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Issue Number:
This is the beginning of changes for #190. It also closes #15. I need to expose the parameter for the git-semver image version to pass in something other than the default `latest`

## Are there any new imports or modules? If so, what are they used for and why?
No.

soda480 commented on vars/edgeXDocker.groovy:
```diff
@@ -135,6 +135,14 @@ def getDockerTags(latest = true, customTags = env.DOCKER_CUSTOM_TAGS) {
+        allTags << 'experimental'
+    }
+
+    if(env.BUILD_STABLE_DOCKER_IMAGE) {
```
can you please add/update `getDockerTags` unit tests to test for both of these conditions, i.e. when BUILD_[EXPERIMENTAL|STABLE]_DOCKER_IMAGE environment variable is set the function returns the expected tags

ernestojeda replied:
See comment from line 138

ernestojeda commented on vars/edgeXDocker.groovy:
```diff
@@ -135,6 +135,14 @@ def getDockerTags(latest = true, customTags = env.DOCKER_CUSTOM_TAGS) {
         allTags << env.SEMVER_BRANCH
     }
 
+    if(env.BUILD_EXPERIMENTAL_DOCKER_IMAGE) {
```
I dont think this will work because Jenkins environment variables all get converted to Strings and because Groovy String based truthy checks only check if the String is null or empty this would still evaluate to true even if the value of the String is the literal word "false". You would have to change this to check the value of the environment variable like:

```Groovy
if(env.BUILD_EXPERIMENTAL_DOCKER_IMAGE == 'true') {
...
```

https://groovyconsole.appspot.com/script/5089484697763840

ernestojeda commented on vars/edgeXBuildGoApp.groovy:
```diff
@@ -296,6 +301,34 @@ def call(config) {
+                }
+                steps {
+                    sshagent (credentials: ['edgex-jenkins-ssh']) {
+                        sh 'echo y | ./scripts/update-named-tag.sh "v${OG_VERSION}" "experimental"'
```
I am not sure this will work properly. This script would have to be available in every repository's workspace where `edgeXBuildGoApp` is used with the `BUILD_*_DOCKER_IMAGE ` environment variables. I rather use a centralized script for this. It may work if we use `libraryResource` in this scenario to load the `update-named-tag.sh` script from edgex-global-pipelines resource dir. We would also have to move that script to the resources folder which I think is not a big deal.

ernestojeda commented on vars/edgeXBuildGoApp.groovy:
```diff
@@ -296,6 +301,34 @@ def call(config) {
+                when {
+                    allOf {
+                        environment name: 'BUILD_EXPERIMENTAL_DOCKER_IMAGE', value: 'true'
+                        expression { edgex.getBranchName() =~ /^master$/ || true } // add || true  to test on sandbox
```
Why not use `edgex.isReleaseStream()`? It is still valid for all repo's except edgex-global-pipelines unless I am mistaken on that.

lranjbar replied:
edgex.isReleaseStream() is a bit too broad. The experimental / stable pattern should be tied to one branch.

ernestojeda commented on vars/edgeXBuildGoApp.groovy:
```diff
@@ -296,6 +301,34 @@ def call(config) {
+                when {
+                    allOf {
+                        environment name: 'BUILD_STABLE_DOCKER_IMAGE', value: 'true'
+                        expression { edgex.getBranchName() =~ /^master$/ || true } // add || true  to test on sandbox
```
Same comment as line 309

ernestojeda commented on vars/edgeXBuildGoApp.groovy:
```diff
@@ -296,6 +301,34 @@ def call(config) {
+                }
+                steps {
+                    sshagent (credentials: ['edgex-jenkins-ssh']) {
+                        sh 'echo y | ./scripts/update-named-tag.sh "v${OG_VERSION}" "stable"'
```
Same comment as line 314

ernestojeda commented on vars/edgex.groovy:
```diff
@@ -175,6 +175,20 @@ def getPreviousCommit(commit) {
+
+// Return true when the commit message follows the pattern "stage(...): ..."
+def isStagingCommit(commit) {    
+    return !!(commit =~ /^stage\(.+\):.+/)
```
`stage` is a bit overloaded. Maybe we could consider something else?

ernestojeda commented:
Another question about control flow I have is how do we rollbacks if an issue is found in stable for instance without having to rebuild the project again preferably.

lranjbar commented:
Update 6/24/20:

- Added more unit tests for edgeXDocker.getDockerTags and updated to string comparison
- Moved the script into the resources folder so it can be loaded by libraryResource(). This call is now encapsulated in a new function edgeXUpdateNamedTag()
- Added unit tests for edgeXUpdateNamedTag()
- Did functional testing using sample-service functional testing PR can be found at:
https://github.com/edgexfoundry/sample-service/pull/42/files

TODO remaining:
- Work out the flow for stable updates on commit messages.

lranjbar commented:
Update: 6/25/20

I added in the functionality for checking the commit message and building a `stable/named` image. Unit tests and functional tests passed in the PR stage of sample-service.

@soda480 @ernestojeda

soda480 commented on vars/edgeXSemver.groovy:
```diff
@@ -48,7 +46,11 @@ def call(command = null, credentials = 'edgex-jenkins-ssh', debug = true) {
                 }
             }
-            semverVersion = sh(script: 'git semver', returnStdout: true).trim()
+
```
If you wish git-semver to set an explicit version as part of init, then the version needs to be specified to the git-semver init command, i.e. the following logic should be added after after the line: `setHeadTagEnv(credentials)`
```
if (semverVersion) {
    semverCommand << "-ver=${semverVersion}"
    semverCommand << "-force"
}
```

ernestojeda commented on vars/edgeXBuildGoApp.groovy:
```diff
@@ -383,13 +439,16 @@ def toEnvironment(config) {
     def _buildSnap           = edgex.defaultFalse(config.buildSnap)
 
+    def _buildExperimentalDockerImage  = edgex.defaultFalse(config.buildExperimentalDockerImage)
+    def _buildStableDockerImage        = edgex.defaultFalse(config.buildStableDockerImage)
```
I dont see the need for this. Since the build commit message should control this.

ernestojeda commented on vars/edgeXBuildGoApp.groovy:
```diff
@@ -55,7 +55,34 @@ def call(config) {
+                        echo("GIT_COMMIT: [$env.GIT_COMMIT]")
+                        echo("commitMsg: [$_commitMsg]")
+
+                        if(edgex.isBuildCommit(_commitMsg)) {
+                            sh "echo This is a build commit."
+                            
+                            def _parsedCommitMsg = edgex.parseBuildCommit(_commitMsg)
+                            def _buildVersion = _parsedCommitMsg.version
+                            def _namedTag = _parsedCommitMsg.namedTag
+                            
+                            echo("buildVersion: [$_buildVersion]")
+                            echo("namedTag: [$_namedTag]")
+
+                            edgeXSemver('init', _buildVersion)  // <-- Generates a VERSION file and .semver directory
+                            env.OG_VERSION = env.VERSION
+                            env.NAMED_TAG = _namedTag
+                            env.BUILD_STABLE_DOCKER_IMAGE = true
+                            sh "echo Archived original version: [$OG_VERSION]"
+                        }
+                        else {
+                            edgeXSemver 'init' // <-- Generates a VERSION file and .semver directory
+                            env.OG_VERSION = env.VERSION
+                            sh "echo Archived original version: [$OG_VERSION]"
+                        }
```
Can we somehow encapsulate this into a function that sets the proper env vars? It would just clean this file up a bit.

lranjbar replied:
I did try this but since you have to call edgeXSemver to get some of the values it doesn't clean up a whole lot. I cleaned up the echo statements as much as possible.

ernestojeda commented:
Also can we rename the `build commit` concept to `bump commit`? build is a bit general So

```
bump(master): [v1.0.0-dev.1,stable] set v1.0.0-dev.1 to stable
```

lranjbar commented:
> Also can we rename the `build commit` concept to `bump commit`? build is a bit general So
> 
> ```
> bump(master): [v1.0.0-dev.1,stable] set v1.0.0-dev.1 to stable
> ```

I disagree with this because it is triggering a rebuild bump implies that it isn't rebuilding anything.

lranjbar commented:
> Here is the control flow as I understand it.
> 
> 1. I create a feature branch and make my proposed change
> 2. I open a PR with new proposed change, (I may iterate on this change experimental is not changed).
> 3. My PR is merged and a semver version of `v1.0.0-dev.1` is created and experimental is now bumped to point to `v1.0.0-dev.1`
> 4. I am ok with this new code being considered `stable` so I create a new branch with an empty commit with a build message.
> 
> ```
> git commit --allow-empty -s -m "build(master): [v1.0.0-dev.1,stable] Bump v1.0.0-dev.1 to stable"
> ```
> 
> 1. Create PR.
> 2. PR is approved and merged.
> 3. During the merge build because this is a `build` commit, semver version is initialized to `v1.0.0-dev.1` everything is rebuilt/tagged/pushed again. At the end of the pipeline, there should not be a need to Semver tag or bump push since `v1.0.0-dev.1` already exists correct? I don't see that in the current code.
> 4. I find an issue with `stable` and need to rollback. So I repeat steps 4-7 with the version I consider `stable`.
> 
> Another concern I have is this:
> 
> I am not sure how many of the EdgeX community devs look at the pipeline stages, but adding two new stages to `edgeXBuildGoApp` that will be unrelated to 99% of all repos may lead to confusion. Maybe we can hide this functionality in one of the existing stages? Maybe during the Semver sub stages.

With the experimental tag it simply happens on every merge. So this is correct.

With the concept of stable, or as I can see it in the other repos a beta or release candidate (rc) it has to trigger a rebuild to get the correct version in the images. As far as the versioned git tag it probably shouldn't make any changes to existing version tags. To skip that part would make sense.

lranjbar commented:
Update 7/13/2020:

Functional testing is finally complete on this one. See the following for a passing build where the images and tags are created and updated properly:
https://jenkins.edgexfoundry.org/blue/organizations/jenkins/edgexfoundry%2Fsample-service/detail/master/30/

@ernestojeda @soda480
